python-package/xgboost/data.py

```diff
@@ -1,553 +1,660 @@
+
+
+def _from_uri(data, missing, feature_names, feature_types):
+    _warn_unused_missing(data, missing)
+    handle = ctypes.c_void_p()
+    _check_call(_LIB.XGDMatrixCreateFromFile(c_str(os_fspath(data)),
+                                             ctypes.c_int(1),
+                                             ctypes.byref(handle)))
+    return handle, feature_names, feature_types
+
+
+def _is_list(data):
+    return isinstance(data, list)
+
+
+def _from_list(data, missing, feature_names, feature_types):
+    raise TypeError('List input data is not supported for data')
+
+
+def _is_tuple(data):
+    return isinstance(data, tuple)
+
+
+def _from_tuple(data, missing, feature_names, feature_types):
+    return _from_list(data, missing, feature_names, feature_types)
+
+
+def _is_iter(data):
+    return isinstance(data, DataIter)
+
+
+def _has_array_protocol(data):
+    return hasattr(data, '__array__')
+
+
+def dispatch_data_backend(data, missing, threads,
+                          feature_names, feature_types):
+    '''Dispatch data for DMatrix.'''
+    if _is_scipy_csr(data):
+        return _from_scipy_csr(data, missing, feature_names, feature_types)
+    if _is_scipy_csc(data):
+        return _from_scipy_csc(data, missing, feature_names, feature_types)
+    if _is_numpy_array(data):
+        return _from_numpy_array(data, missing, threads, feature_names,
+                                 feature_types)
+    if _is_uri(data):
+        return _from_uri(data, missing, feature_names, feature_types)
+    if _is_list(data):
+        return _from_list(data, missing, feature_names, feature_types)
+    if _is_tuple(data):
+        return _from_tuple(data, missing, feature_names, feature_types)
+    if _is_pandas_df(data):
+        return _from_pandas_df(data, missing, threads,
+                               feature_names, feature_types)
+    if _is_pandas_series(data):
+        return _from_pandas_series(data, missing, threads, feature_names,
+                                   feature_types)
+    if _is_cudf_df(data):
+        return _from_cudf_df(data, missing, threads, feature_names,
+                             feature_types)
+    if _is_cudf_ser(data):
+        return _from_cudf_df(data, missing, threads, feature_names,
+                             feature_types)
+    if _is_cupy_array(data):
+        return _from_cupy_array(data, missing, threads, feature_names,
+                                feature_types)
+    if _is_cupy_csr(data):
+        raise TypeError('cupyx CSR is not supported yet.')
+    if _is_cupy_csc(data):
+        raise TypeError('cupyx CSC is not supported yet.')
+    if _is_dlpack(data):
+        return _from_dlpack(data, missing, threads, feature_names,
+                            feature_types)
+    if _is_dt_df(data):
+        _warn_unused_missing(data, missing)
+        return _from_dt_df(data, missing, threads, feature_names,
+                           feature_types)
+    if _has_array_protocol(data):
```

Comment:
How do we handle data types with `__array__` protocol? Does it get converted into a NumPy array?

Reply:
Not really, I can't push a cupy array to numpy array.  It's called `array_like`, we need to add extra dispatching logic for it.  I don't want to do it in this PR.

Reply:
Got it. Do you plan to write another PR for dispatching array-like data?

Reply:
@hcho3 Will do it later.  It's required for passing scikit-learn validations.